src/lib/table/table-data-source.ts

```diff
@@ -91,14 +91,24 @@ export class MatTableDataSource<T> implements DataSource<T> {
   }
 
   /**
-   * Transforms data objects into a filter term that will be used to check against the filter if
-   * a filter is set. By default, the function will iterate over the values of the data object
-   * and convert them to a lowercase string.
-   * @param data Data object to convert to a string that checked for containing the filter term.
+   * Checks if a data object matches the data source's filter string. By default, each data object
+   * is converted to a string of its properties and returns true if the filter has
+   * at least one occurrence in that string. By default, the filter string has its whitespace
+   * removed and the match is case-insensitive. May be overriden for a custom implementation of
+   * filter matching.
+   * @param data Data object used to check against the filter.
+   * @param filter Filter string that has been set on the data source.
+   * @returns Whether the filter matches against the data
    */
-  filterTermAccessor: ((data: T) => string) = (data: T): string => {
+  filterMatcher: ((data: T, filter: string) => boolean) = (data: T, filter: string): boolean => {
+    // Transform the data into a lowercase string of all property values.
     const accumulator = (currentTerm, key) => currentTerm + data[key];
-    return Object.keys(data).reduce(accumulator, '').toLowerCase();
+    const dataStr = Object.keys(data).reduce(accumulator, '').toLowerCase();
```

Comment:
You want to filter on the _keys_?

Reply:
Looks like it, but its actually iterating over each of the object's keys and then reducing it via the accumulator function. That function takes each key and accesses the data's property and appends it to the accumulated value.

In the end an object like `{name: 'John', color: 'Blue', age: '27'}` ends up looking like `'JohnBlue27'`

Would be nice to use `Object.values()` but its not supported by IE